cpp/src/arrow/compute/kernels/hash_aggregate.cc

```diff
@@ -3319,9 +3324,401 @@ struct GroupedListFactory {
+      return Status::OK();
+    }
+    if (transform) {
+      ARROW_ASSIGN_OR_RAISE(other_column, transform(other_column));
+    }
+    DCHECK_EQ(num_groups_, other_column->length());
+    if (!*column) {
+      *column = other_column;
+      return Status::OK();
+    }
+    if ((*column)->null_count() == (*column)->length()) {
+      *column = other_column;
+      return Status::OK();
+    }
+    int64_t expected_non_nulls = (num_groups_ - (*column)->null_count()) +
+                                 (num_groups_ - other_column->null_count());
+    ARROW_ASSIGN_OR_RAISE(auto coalesced,
+                          CallFunction("coalesce", {*column, other_column}, ctx_));
+    // Check that all non-null values in other_column and column were kept in the result.
+    if (expected_non_nulls != num_groups_ - coalesced.null_count()) {
+      DCHECK_GT(expected_non_nulls, num_groups_ - coalesced.null_count());
+      return DuplicateValue();
+    }
+    *column = coalesced.make_array();
+    return Status::OK();
+  }
+
+  Status DuplicateValue() {
+    return Status::Invalid(
+        "Encountered more than one non-null value for the same grouped pivot key");
+  }
+
+  ExecContext* ctx_;
+  std::shared_ptr<DataType> value_type_;
+  int num_keys_;
+  int64_t num_groups_;
+  ArrayVector columns_;
+  // A persistent scratch buffer to store the take indices in Consume
+  BufferBuilder scratch_buffer_;
+};
+
+struct GroupedPivotImpl : public GroupedAggregator {
+  Status Init(ExecContext* ctx, const KernelInitArgs& args) override {
+    DCHECK_EQ(args.inputs.size(), 3);
+    key_type_ = args.inputs[0].GetSharedPtr();
+    options_ = checked_cast<const PivotWiderOptions*>(args.options);
+    DCHECK_NE(options_, nullptr);
+    auto value_type = args.inputs[1].GetSharedPtr();
+    FieldVector fields;
+    fields.reserve(options_->key_names.size());
+    for (const auto& key_name : options_->key_names) {
+      fields.push_back(field(key_name, value_type));
+    }
+    out_type_ = struct_(std::move(fields));
+    out_struct_type_ = checked_cast<const StructType*>(out_type_.get());
+    ARROW_ASSIGN_OR_RAISE(key_mapper_, PivotWiderKeyMapper::Make(*key_type_, options_));
+    RETURN_NOT_OK(accumulator_.Init(ctx, value_type, options_));
+    return Status::OK();
+  }
+
+  Status Resize(int64_t new_num_groups) override {
+    num_groups_ = new_num_groups;
+    return accumulator_.Resize(new_num_groups);
+  }
+
+  Status Merge(GroupedAggregator&& raw_other,
+               const ArrayData& group_id_mapping) override {
+    auto other = checked_cast<GroupedPivotImpl*>(&raw_other);
+    return accumulator_.Merge(std::move(other->accumulator_), group_id_mapping);
+  }
+
+  Status Consume(const ExecSpan& batch) override {
+    DCHECK_EQ(batch.values.size(), 3);
+    auto groups = batch[2].array.GetSpan<const uint32_t>(1, batch.length);
+    if (!batch[1].is_array()) {
+      return Status::NotImplemented("Consuming scalar pivot value");
```

Comment:
Does it make sense to support scalar here as well? (similar to the non hashed version)